src/server/project.ts

```diff
@@ -1955,6 +1955,10 @@ namespace ts.server {
             return super.updateGraph();
         }
 
+        hasRoots() {
+            return !!ts.some(this.rootFileNames);

```

Comment:
shouldnt thios be always true instead? Since this project will always have roots ?

Reply:
checking rootFileNames seems incorrect since during markAsDirty it is set to undefined to recalculate?

Reply:
Oh I see, Changing hasRoots seems confusing though... why not directly check `this.autoImportProviderHost.rootFileNames` instead
```
updateProjectIfDirty(this.autoImportProviderHost);
                if (!this.autoImportProviderHost.hasRoots()) {
                    this.autoImportProviderHost.close();
```

Reply:
No; if the main project imports something that was previously a root in here, that root “moves” from here to the main project, which could render this project empty after `updateGraph`, at which point the main project will close this project. So it doesn’t last very long without roots, but being in that state is what causes it to be closed.

Reply:
That would be ok